Fix auto-objects and make rw_ok assumptions create proper backing objects

tautschnig opened the pull request:
Fix the auto-objects mechanism and extend `__CPROVER_rw_ok` in assumptions to create properly-typed backing objects, enabling verification of code that operates on arrays and inductive data structures (linked lists, trees) through nondet pointers.

### Changes

**Commit 1: Fix goto-symex' auto-objects feature**
- Fix `trigger_auto_object` name check (`auto_object` vs `symex::auto_object`)
- Move trigger before dereference so it operates on the pointer, not the resolved object
- Fix the "done already?" check for initial L2 index

**Commit 2: Enforce equal object_size for same-object pointers**
- Add pairwise constraints in `finish_eager_conversion()` so that `object_size` propagates correctly through pointer aliases (e.g., `b = a` followed by `rw_ok(a, size)`)

**Commit 3: Create array-typed objects for rw_ok and support inductive data structures**
- When `__CPROVER_assume(__CPROVER_rw_ok(ptr, size))` is processed and `size` is a compile-time constant larger than `sizeof(*ptr)`, create an array object (like `malloc` does)
- For structs with pointer members, recursively initialize pointer members via auto-objects, creating lazy chains bounded by `--unwind`
- Restrict auto-object initialization to pointers explicitly assumed valid via `rw_ok` (soundness safeguard)
- Add comprehensive user documentation in `doc/cprover-manual/memory-primitives.md`

### What users can now do

```c
// Arrays via rw_ok
unsigned int *a;
__CPROVER_assume(__CPROVER_rw_ok(a, 3 * sizeof(*a)));
a[0] = 10; a[1] = 20; a[2] = 30;
assert(a[0] == 10); // succeeds

// Linked lists via rw_ok + --unwind
struct node { int data; struct node *next; };
struct node *head;
__CPROVER_assume(__CPROVER_rw_ok(head, sizeof(*head)));
__CPROVER_assume(head != 0);
// head is now a valid linked list bounded by --unwind
```

### Soundness

Nondet pointers without `rw_ok` assumptions still fail all pointer checks. This is enforced by tracking which pointers had `rw_ok` assumed and only triggering auto-object initialization for those. A regression test (`r_w_ok_no_false_negative`) verifies this.

Fixes: #7829
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Fix goto-symex' auto-objects feature

The auto-objects feature creates fresh objects for pointers whose
symbol name starts with auto_object. It was broken in two ways:

1. trigger_auto_object checked for base names starting with
   "symex::auto_object", but get_fresh_aux_symbol produces base
   names starting with "auto_object" (without the symex:: prefix).
   Fix by also matching the "auto_object" prefix.

2. trigger_auto_object was called after dereference, operating on
   the resolved (failed) symbol rather than the pointer itself.
   Since the failed symbol is the pointed-to object (not a pointer),
   initialize_auto_object had no effect. Move the call to before
   dereference so it operates on the pointer and can assign it to
   point to a fresh auto-object.

3. The "done already" check used has_key which doesn't account for
   the initial L2 index. Use find and check for index <= 1 instead.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Fix goto-symex' auto-objects feature

The auto-objects feature creates fresh objects for pointers whose
symbol name starts with auto_object. It was broken in two ways:

1. trigger_auto_object checked for base names starting with
   "symex::auto_object", but get_fresh_aux_symbol produces base
   names starting with "auto_object" (without the symex:: prefix).
   Fix by also matching the "auto_object" prefix.

2. trigger_auto_object was called after dereference, operating on
   the resolved (failed) symbol rather than the pointer itself.
   Since the failed symbol is the pointed-to object (not a pointer),
   initialize_auto_object had no effect. Move the call to before
   dereference so it operates on the pointer and can assign it to
   point to a fresh auto-object.

3. The "done already" check used has_key which doesn't account for
   the initial L2 index. Use find and check for index <= 1 instead.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

martin-cs reviewed (approved):
Yes but we do have to be a little careful of the "don't give semantics to undefined behaviour" rule.

kiro-agent committed:
Create backing objects for rw_ok and support inductive data structures

When __CPROVER_assume(__CPROVER_rw_ok(ptr, size)) is processed during
symex and size is a compile-time constant, create a concrete backing
object and constrain the pointer to point to it. For multi-element
sizes (size > sizeof(*ptr)), use symex_assign to set the pointer to
the array object. For single-element sizes, add the object to the
value set and assume ptr == &obj, preserving pre-existing aliases.

For structs with pointer members, recursively initialize pointer
members via auto-objects when the failed symbol is first dereferenced,
creating lazy chains bounded by --unwind. Restrict auto-object
initialization to pointers explicitly assumed valid via rw_ok
(soundness safeguard).

Remove no-new-smt exclusion tag from r_w_ok_auto_object_array since
the test uses __CPROVER_forall which is unsupported by the
incremental SMT2 backend. Tag it broken-smt-backend as all SMT
backends fail on the array_equal/forall assertions.

Add comprehensive user documentation in doc/cprover-manual/memory-primitives.md.

Fixes: #7829

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

kiro-agent committed:
Create backing objects for rw_ok and support inductive data structures

When __CPROVER_assume(__CPROVER_rw_ok(ptr, size)) is processed during
symex and size is a compile-time constant, create a concrete backing
object and constrain the pointer to point to it. For multi-element
sizes (size > sizeof(*ptr)), use symex_assign to set the pointer to the
array object so that whole-array operations resolve to a single target.
For single-element sizes, add the object to the value set and assume
ptr == &obj, which preserves pre-existing aliases.

Only rw_ok expressions that are top-level conjuncts of the assumption
drive object creation; an rw_ok nested under a negation, disjunction or
conditional is ignored, so e.g. !rw_ok(p, ...) does not silently make
the path infeasible and p == &g || rw_ok(p, ...) keeps the p == &g
disjunct. The logic lives in the helper
goto_symext::try_create_rw_ok_backing_object. If the constant size is
smaller than one element, a warning is emitted and no object is created.

For structs with pointer members, recursively initialize pointer
members via auto-objects when the failed symbol is first dereferenced,
creating lazy chains bounded by --unwind. Restrict auto-object
initialization to pointers explicitly assumed valid via rw_ok
(soundness safeguard).

r_w_ok_auto_object_array is tagged broken-smt-backend and
broken-cprover-smt-backend (the array_equal/forall assertions are
unsupported by those backends) and keeps no-new-smt, as the incremental
SMT2 backend does not support __CPROVER_forall.

Add user documentation in doc/cprover-manual/memory-primitives.md,
including limitations (non-constant and sub-element sizes, the single-
vs multi-element alias asymmetry, and chained dereference expressions
through freshly-created inductive nodes), and regression tests for the
array out-of-bounds check, tree traversal, negated and disjunctive
rw_ok, deep reads via intermediate variables, and the sub-element
warning.

Fixes: #7829

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Lift nested dereferences into temporaries

Add a goto-program transformation that rewrites a chained dereference such
as `a->b->c` into

    tmp = a->b;
    ... tmp->c ...

introducing a fresh local temporary for the result of the inner dereference.
The pass is a small forward common-subexpression elimination over the
"intermediate pointers" of chained dereferences: each such pointer is hoisted
to a temporary at its first use (including its use in a guard) and the
temporary is reused at all subsequent uses until a write might invalidate it.
Invalidation uses a dirtyt (address-taken) analysis so that writes to
unrelated, non-address-taken variables do not discard available temporaries.
Jumps to a rewritten instruction are preserved via insert_before_swap.

This has two benefits:

* The same pointer is no longer dereferenced more than once for a chained
  access (chained dereferences are otherwise re-evaluated for each use).
* The intermediate pointer becomes an ordinary local variable with its own
  object identity. Downstream analyses (notably the value-set dereference and
  auto-object initialisation in goto-symex) then treat it like a hand-written
  intermediate variable. In particular this makes a guarded chained read such
  as `if(p->next) p->next->data` verify when `p` was assumed valid via
  __CPROVER_rw_ok: previously the intermediate pointer was read (and lazily
  materialised) twice, so the guard and the dereference disagreed and the
  access was reported as a spurious failure.

The transformation is conservatively skipped for concurrent programs (those
containing START_THREAD): reusing a single read of a possibly-shared
intermediate pointer would collapse two reads of shared memory into one,
removing thread interleavings and potentially hiding data-race-dependent bugs.

The pass runs in process_goto_program and is on by default; it can be
disabled with --no-lift-nested-dereferences. Regression test
lift_nested_dereferences1 pins both the on (verifies) and off (spurious
failure) behaviour, and a unit test covers the single-threaded, plain-pointer,
and concurrent cases. The r_w_ok_no_false_negative descriptor is made robust
to the introduced temporary name.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Document and test chained dereferences for rw_ok inductive structures

With the nested-dereference lifting pass, chained dereferences such as
head->next->data now verify directly for structures assumed valid via
__CPROVER_rw_ok, so the previous "Chained dereference expressions" limitation
no longer applies. Remove it from doc/cprover-manual/memory-primitives.md and
add a note that chained dereferences can be used directly.

Rewrite the r_w_ok_deep_read regression to read the deep scalar fields through
inline chained dereferences (head->next->data, head->next->next->data) rather
than via explicit intermediate variables, exercising the lifted path.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>
